lib/bankster/bic.ex

```diff
@@ -1,23 +1,148 @@
 defmodule Bankster.Bic do
   @moduledoc """
   Module provides some SWIFT BIC related functions.
+
+  See: https://en.wikipedia.org/wiki/ISO_9362
   """
 
-  ## -- Module attributes
-  @bic_validation_regex ~r/^([a-zA-Z]){4}([a-zA-Z]){2}([0-9a-zA-Z]){2}([0-9a-zA-Z]{3})?$/
+  defstruct [:bank, :country, :location, :branch]
+
+  @type t :: %__MODULE__{
+          bank: binary(),
+          country: binary(),
+          location: binary(),
+          branch: binary() | nil
+        }
+
+  @bic_regex ~r/^(?<bank>[a-zA-Z]{4})\s*(?<country>[a-zA-Z]{2})\s*(?<location>[0-9a-zA-Z]{2})\s*(?<branch>[0-9a-zA-Z]{3})?$/
 
-  ## -- Functions
   @doc """
-  Validates a given string whether it's a valid SWIFT BIC
+  Parses a BIC string into its components.
+
+  ## Examples
+    iex> Bankster.Bic.new("INVALIDBIC")
+    :error
+
+    iex> Bankster.Bic.new("AAAABBCC123")
+    %Bankster.Bic{bank: "AAAA", country: "BB", location: "CC", branch: "123"}
+
+    iex> Bankster.Bic.new("AAAABBCC")
+    %Bankster.Bic{bank: "AAAA", country: "BB", location: "CC", branch: nil}
+
+    iex> Bankster.Bic.new("AAAA BB CC 123")
+    %Bankster.Bic{bank: "AAAA", country: "BB", location: "CC", branch: "123"}
+
+    iex> Bankster.Bic.new("AAAA BB CC")
+    %Bankster.Bic{bank: "AAAA", country: "BB", location: "CC", branch: nil}
+
+    iex> Bankster.Bic.new("AAAA BB CC     123")
+    %Bankster.Bic{bank: "AAAA", country: "BB", location: "CC", branch: "123"}
+  """
+  @spec new(binary()) :: {:ok, t()} | :error
```

Comment:
This _could_ probably be defined as

```
@spec new(term()) :: {:ok, t()} | :error
```

Since it has the fall through `new(_)` definition.

Reply:
Is `new` a good name for that function, as it _parses_ a given BIC into its components? In my opinion `parse(bic)` might be a better name.

I think two parsing functions would be nice.

```elixir
@spec parse(term()) :: {:ok, t()} | :error
```

The parsing functions, which raises an error, can be used in pipelines.
```elixir
@spec parse!(term()) :: t()
```